### PR TITLE
Handle CPP

### PR DESCRIPTION
**Pull request by tfausak**
I should have expected this problem while working on #23. Using GHC as a library doesn't handle CPP in the way you might expect. Since CPP is an external program, GHC appears to identify the language extension and change how it parses, but it doesn't actually succeed. It seemed to me like the easiest way out was to use `cpphs` to run CPP when necessary. A couple things though:

- It's not clear when to run CPP. GHC needs to semi-parse the file to find language extensions. Are you supposed to run CPP before or after that? After would make sense because you can do `{-# language CPP #-}`, but the language pragmas themselves could be behind CPP so who knows. Some experimentation is in order.

- Various CPP options from the Cabal file need to be passed through. Any defined variables need to be set up. And any actual options need to be parsed and applied, which might be tricky. Not every package uses (or can use?) `cpphs`, so options may need to be converted from GCC or Clang into `cpphs`. Sounds like fun. 

In spite of the above caveats, I feel reasonably confident saying that many of the parse failures will be fixed by these changes.

### CODE REVIEW

**Comment by tfausak** [comment]
Before:
- 1,520,168 modules
- 1,188,414 (78%) passed
- 331,754 (22%) failed

After:
- 1,521,759 modules
- 1,224,409 (80%) passed
- 297,350 (20%) failed

So ... better, but not as dramatic as I expected. Time to start actually digging into the failed parses to see why they're not working. 

One thing I'll need to consider is the `default-language` field in the Cabal file. Switching between Haskell98 and Haskell2010 changes the default extensions.

**Comment by tfausak** [comment]
This section of the GHC user guide might come in handy: https://downloads.haskell.org/~ghc/8.10.2/docs/html/users_guide/phases.html#options-affecting-the-c-pre-processor

**Comment by tfausak** [comment]
Ah, I think I see why there were so many failures. I checked for the CPP extension from the list of extensions passed in via Cabal, but I didn't check for the CPP pragma. The most recent commit fixes this problem. 

>  language pragmas themselves could be behind CPP

Turns out you can't disable CPP with `{-# language NoCPP #-}`. Once it's on, it's on. And you can't hide it behind some CPP conditional because CPP isn't enabled until it finds that pragma. You could set `default-extensions: CPP` and then do `#if 1\n{#- language NoCPP #-}\n#endif` but that wouldn't do anything since you can't disable CPP via a pragma. 

> CPP options from the Cabal file need to be passed

Hopefully most files with CPP will work even without doing this, but I'm less confident now than I was before.

**Comment by tfausak** [comment]
Apparently my GHC parsing setup doesn't handle Literate Haskell, so I'll have to figure that out. Fortunately there aren't that many `*.lhs` files:

``` sql
sqlite> select count(*) from exposed_modules where file like '%.lhs' ;
7816
```

**Comment by tfausak** [comment]
I think another big chunk of this is going to be figuring out which extensions are necessary to emulate older versions of GHC. For example I think `MagicHash` used to effectively always be enabled until it was squirreled away under a language extension. Perhaps this changed around GHC 6.8.1? https://downloads.haskell.org/ghc/8.10.2/docs/html/users_guide/glasgow_exts.html#extension-MagicHash

**Comment by tfausak** [comment]
Ooh running this ran into my first crash while trying to parse a source file:

```
2020-08-19T01:45:04.262Z #error "unsupported sizeof(HsInt)"
in c/NewBinary/0.1.1/NewBinary/Binary.hs  at line 594 col 1
CallStack (from HasCallStack):
  error, called at ./Language/Preprocessor/Cpphs/CppIfdef.hs:113:21 in cpphs-1.20.9.1-qNwIEc8el4KR2GEETsFCT:Language.Preprocessor.Cpphs.CppIfdef
```

I need to capture errors from `cpphs` as well.

**Comment by tfausak** [comment]
Ugh, of course. `cpphs` runs in `IO`, but it also throws impure exceptions via `error`. Since I wasn't forcing the result, the `IO` action completed and my exception handler never got called. Fortunately it's easy enough to fix.

**Comment by tfausak** [comment]
With these changes the new numbers are:

- Total: 1,521,885
- Pass: 1,371,436 (90%)
- Fail: 150,449 (10%)

So that's a huge improvement, but still a lot of failures.

**Comment by tfausak** [comment]
Huh, this is interesting. The list of language extensions passed into GHC doesn't seem to enable the implied extensions. For example:

``` hs
-- fails
Ghc.parse
  [ ( True, Ext.RankNTypes ) ]
  ""
  "f :: forall a . a -> a\nf a = a"

-- succeeds
Ghc.parse
  []
  ""
  "{-# language RankNTypes #-}\nf :: forall a . a -> a\nf a = a"
```

At first this didn't make sense to me, but apparently it's possible to enable an extension and then disable an extension that it implies. For example, `{-# language RankNTypes, NoExplicitForAll #-}`. Not sure why exactly you'd want to do that. 

Regardless what I need to do is figure out which extensions are implied by other extensions and then enable those too.

**Comment by tfausak** [comment]
The logic for implied language extensions does appear in the GHC source tree, but I can't seem to find it in my installed version's module hierarchy. It's called `impliedXFlags` and it's in the `GHC.Driver.Session` module. 

https://gitlab.haskell.org/ghc/ghc/-/blob/a1f34d37b47826e86343e368a5c00f1a4b1f2bce/compiler/GHC/Driver/Session.hs#L3960

It seems to be the result of a recent merge request? https://gitlab.haskell.org/ghc/ghc/-/merge_requests/2740

That belongs to a much larger issue: https://gitlab.haskell.org/ghc/ghc/-/issues/13009

There's also a proposal: https://github.com/ghc-proposals/ghc-proposals/pull/57

And a wiki page: https://gitlab.haskell.org/ghc/ghc/-/wikis/Make-GHC-codebase-more-modular

TL;DR: Doesn't seem to be available currently. I should probably just copy-paste the logic.

**Comment by tfausak** [comment]
``` sql
sqlite> select count(*) from exposed_modules ;
1522140
sqlite> select count(*) from exposed_modules where parsed = true ;
1392693 -- 91%
sqlite> select count(*) from exposed_modules where parsed = false ;
129447 -- 9%
```

**Comment by tfausak** [comment]
I was curious how this shook out per file extension:

Extension | Failed parses
--- | ---
none | 52910
`.chs` | 33514
`.hs` | 18721
`.hsc` | 14065
`.lhs` | 7816
`.x` | 1749
`.y` | 586
`.cpphs` | 59
`.xpphs` | 20
`.gc` | 7
`.xhs` | 0

So the biggest problem is simply not finding the module source. Then `.chs` and `.hsc` contribute a bunch.

**Comment by tfausak** [comment]
Clearly I still have a lot of work to do, but I think it's a good time to move on to something else. There's always going to be _something_ I can't parse since I'm not running TH or arbitrary build tools like `markdown-unlit`. That means my code is going to have to deal with that case. I can't count on 100% success. 

I'm sure there's still a bunch of low hanging fruit, like figuring out language extensions for older versions of GHC or correctly dealing with LHS. Regardless I think it's a better use of my time to move on to finding exported identifiers and parsing their documentation.

**Comment by tfausak** [comment]
Some breadcrumbs for handling other file types:

- `.chs`: https://hackage.haskell.org/package/c2hs
- `.hsc`: https://hackage.haskell.org/package/hsc2hs
- `.lhs`: I'm having trouble finding information about this one. I think internally GHC uses a C program called `unlit` for this, but it's not on my PATH. As far as I can tell GHC doesn't expose a library function for doing this step. The package I'm using for CPP (`cpphs`) does have a function for this. Perhaps I can just use that? https://hackage.haskell.org/package/cpphs-1.20.9.1/docs/Language-Preprocessor-Unlit.html#v:unlit
- `.x`: https://www.haskell.org/alex/
- `.y`: https://www.haskell.org/happy/
- `.cpphs`: Is this just a Haskell file with the CPP language extension enabled? Seems like it. 
  - Related note: I'll probably have to resolve `#include`s to really support CPP. 
- `.xpphs`: These are only used by one package: `Hs2lib`. Seems like they're basically the same as `.cpphs`. Not sure what the difference is.
